hw02-DIYList/src/main/java/ru/otus/hw02/DIYList.java

```diff
@@ -0,0 +1,216 @@
+package ru.otus.hw02;
+
+import java.util.*;
+
+public class DIYList<E> implements List<E> {
+
+    private Object[] elements = new Object[10];
+    private int size = 0;
+
+    private class ListItr implements ListIterator<E> {
+
+        private  int cursor = 0;
+        private int lastRet = -1;
+
+        public boolean hasPrevious() {
+            return cursor != 0;
+        }
+
+        public E next() {
+            if (cursor >= size)
+                throw new NoSuchElementException();
+            cursor++;
+            return (E) elements[lastRet = cursor - 1];
+        }
+
+        public void remove() {
+            if (lastRet < 0)
+                throw new IllegalStateException();
+
+            DIYList.this.remove(lastRet);
+            cursor = lastRet;
+            lastRet = -1;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return cursor != size;
+        }
+
+        public int nextIndex() {
+            return cursor;
+        }
+
+        public int previousIndex() {
+            return cursor - 1;
+        }
+
+        public E previous() {
+            if ( cursor - 1 < 0 ) {
+                throw new NoSuchElementException();
+            }
+            return (E) elements[lastRet = cursor - 1];
+        }
+
+        public void set(E e) {
+                DIYList.this.set(cursor - 1, e);
+        }
+
+        public void add(E e) {
+            DIYList.this.add(cursor, e);
+            cursor++;
+            lastRet = -1;
+        }
+    }
+
+    private void indexCheck(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void checkSize() {
+        if (size == elements.length) {
+            elements = Arrays.copyOf(elements, elements.length * 2);
+        }
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return size == 0;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        if (size == 0) return false;
+        if (o == null) {
+            for (int i = 0; i < size; i++) {
+                if (elements[i] == null) {
+                    return true;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (o.equals(elements[i])) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        return new ListItr();
+    }
+
+    @Override
+    public Object[] toArray() {
+        return Arrays.copyOf(elements, size);
+    }
+
+    @Override
+    public <T> T[] toArray(T[] a) {
+        return null;
+    }
+
+    @Override
+    public boolean add(E e) {
+        elements[size++] = e;
+        checkSize();
+        return true;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        return false;
```

Comment:
В не реализованных методах лучше бросать исключение.
```suggestion
        throw new UnsupportedOperationException();
```

Чтобы внешний код сразу понял, что что-то не так.
При успешном завершении метода во внешнем коде можно и не понять, что на самом деле реализации не было, и это может уже всплыть где-то в самом неожиданном месте.